File: src/forge/core/models/manufacturing/enums.py
"""Shared enumerations for manufacturing domain models.

These enums define Forge-canonical values for status, type, and
classification fields. Adapters map source-system-specific values
(e.g., WMS EnumBarrelDisposition, MES BatchStatus) to these canonical
enums during ingestion.
"""

from enum import StrEnum

# ── Manufacturing Unit ──────────────────────────────────────────────


class UnitStatus(StrEnum):
    """Lifecycle status of a manufacturing unit (barrel, batch, tank)."""

    PENDING = "PENDING"  # Created but not yet active
    ACTIVE = "ACTIVE"  # In production or storage
    COMPLETE = "COMPLETE"  # Finished its lifecycle
    HELD = "HELD"  # On quality or regulatory hold
    SCRAPPED = "SCRAPPED"  # Disposed or dumped
    TRANSFERRED = "TRANSFERRED"  # Moved to external system


class LifecycleState(StrEnum):
    """Granular lifecycle phase within a manufacturing unit's journey.

    Derived from ISA-88 batch states and WMS barrel disposition.
    """

    CREATED = "CREATED"
    FILLING = "FILLING"
    IN_PROCESS = "IN_PROCESS"
    AGING = "AGING"  # Whiskey-specific but generalizable to curing/resting
    IN_STORAGE = "IN_STORAGE"
    IN_TRANSIT = "IN_TRANSIT"
    SAMPLING = "SAMPLING"
    WITHDRAWN = "WITHDRAWN"
    DUMPED = "DUMPED"
    COMPLETE = "COMPLETE"


# ── Physical Asset ──────────────────────────────────────────────────


class AssetType(StrEnum):
    """Type of physical asset in the manufacturing hierarchy.

    Follows ISA-95 equipment hierarchy levels where applicable.
    """

    ENTERPRISE = "ENTERPRISE"
    SITE = "SITE"
    AREA = "AREA"
    WORK_CENTER = "WORK_CENTER"  # ISA-95 "Work Center"
    WORK_UNIT = "WORK_UNIT"  # ISA-95 "Work Unit"
    STORAGE_ZONE = "STORAGE_ZONE"  # Warehouse floor/section
    STORAGE_POSITION = "STORAGE_POSITION"  # Individual rack/position
    EQUIPMENT = "EQUIPMENT"  # Fermenter, still, tank, etc.
    STAGING_AREA = "STAGING_AREA"  # Holding/receiving area


class AssetOperationalState(StrEnum):
    """Current operational state of an asset."""

    IDLE = "IDLE"
    RUNNING = "RUNNING"
    MAINTENANCE = "MAINTENANCE"
    CHANGEOVER = "CHANGEOVER"
    OFFLINE = "OFFLINE"
    FAULTED = "FAULTED"


# ── Operational Event ───────────────────────────────────────────────


class EventSeverity(StrEnum):
    """Severity level for operational events."""

    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class EventCategory(StrEnum):
    """High-level category for operational events."""

    PRODUCTION = "PRODUCTION"
    QUALITY = "QUALITY"
    LOGISTICS = "LOGISTICS"
    MAINTENANCE = "MAINTENANCE"
    SAFETY = "SAFETY"
    COMPLIANCE = "COMPLIANCE"


# ── Business Entity ─────────────────────────────────────────────────


class EntityType(StrEnum):
    """Type of business entity."""

    CUSTOMER = "CUSTOMER"
    VENDOR = "VENDOR"
    PARTNER = "PARTNER"
    INTERNAL = "INTERNAL"


# ── Work Order ──────────────────────────────────────────────────────


class WorkOrderStatus(StrEnum):
    """Status of a work order."""

    DRAFT = "DRAFT"
    PENDING = "PENDING"
    SCHEDULED = "SCHEDULED"
    IN_PROGRESS = "IN_PROGRESS"
    PAUSED = "PAUSED"
    COMPLETE = "COMPLETE"
    CANCELLED = "CANCELLED"


class WorkOrderPriority(StrEnum):
    """Priority level for work orders."""

    LOW = "LOW"
    NORMAL = "NORMAL"
    HIGH = "HIGH"
    URGENT = "URGENT"


# ── Production Order ────────────────────────────────────────────────


class OrderStatus(StrEnum):
    """Status of a production order."""

    DRAFT = "DRAFT"
    PLANNED = "PLANNED"
    RELEASED = "RELEASED"
    IN_PROGRESS = "IN_PROGRESS"
    PAUSED = "PAUSED"
    COMPLETE = "COMPLETE"
    CLOSED = "CLOSED"
    CANCELLED = "CANCELLED"


# ── Quality ─────────────────────────────────────────────────────────


class SampleOutcome(StrEnum):
    """Result of a quality sample evaluation."""

    PASS = "PASS"
    FAIL = "FAIL"
    INCONCLUSIVE = "INCONCLUSIVE"
    PENDING = "PENDING"

File: go/internal/models/manufacturing/enums.go
// Package manufacturing defines canonical manufacturing domain entities.
//
// These types are the Go equivalents of the Python models in
// src/forge/core/models/manufacturing/. Adapters map source-system-specific
// values to these canonical types during ingestion.
package manufacturing

// -- Manufacturing Unit --

// UnitStatus is the lifecycle status of a manufacturing unit.
type UnitStatus string

const (
	UnitStatusPending     UnitStatus = "PENDING"
	UnitStatusActive      UnitStatus = "ACTIVE"
	UnitStatusComplete    UnitStatus = "COMPLETE"
	UnitStatusHeld        UnitStatus = "HELD"
	UnitStatusScrapped    UnitStatus = "SCRAPPED"
	UnitStatusTransferred UnitStatus = "TRANSFERRED"
)

// LifecycleState is a granular lifecycle phase within a manufacturing unit's journey.
type LifecycleState string

const (
	LifecycleCreated   LifecycleState = "CREATED"
	LifecycleFilling   LifecycleState = "FILLING"
	LifecycleInProcess LifecycleState = "IN_PROCESS"
	LifecycleAging     LifecycleState = "AGING"
	LifecycleInStorage LifecycleState = "IN_STORAGE"
	LifecycleInTransit LifecycleState = "IN_TRANSIT"
	LifecycleSampling  LifecycleState = "SAMPLING"
	LifecycleWithdrawn LifecycleState = "WITHDRAWN"
	LifecycleDumped    LifecycleState = "DUMPED"
	LifecycleComplete  LifecycleState = "COMPLETE"
)

// -- Physical Asset --

// AssetType classifies physical assets in the ISA-95 hierarchy.
type AssetType string

const (
	AssetTypeEnterprise      AssetType = "ENTERPRISE"
	AssetTypeSite            AssetType = "SITE"
	AssetTypeArea            AssetType = "AREA"
	AssetTypeWorkCenter      AssetType = "WORK_CENTER"
	AssetTypeWorkUnit        AssetType = "WORK_UNIT"
	AssetTypeStorageZone     AssetType = "STORAGE_ZONE"
	AssetTypeStoragePosition AssetType = "STORAGE_POSITION"
	AssetTypeEquipment       AssetType = "EQUIPMENT"
	AssetTypeStagingArea     AssetType = "STAGING_AREA"
)

// AssetOperationalState is the current operational state of an asset.
type AssetOperationalState string

const (
	AssetStateIdle        AssetOperationalState = "IDLE"
	AssetStateRunning     AssetOperationalState = "RUNNING"
	AssetStateMaintenance AssetOperationalState = "MAINTENANCE"
	AssetStateChangeover  AssetOperationalState = "CHANGEOVER"
	AssetStateOffline     AssetOperationalState = "OFFLINE"
	AssetStateFaulted     AssetOperationalState = "FAULTED"
)

// -- Operational Event --

// EventSeverity indicates event severity level.
type EventSeverity string

const (
	EventSeverityInfo     EventSeverity = "INFO"
	EventSeverityWarning  EventSeverity = "WARNING"
	EventSeverityError    EventSeverity = "ERROR"
	EventSeverityCritical EventSeverity = "CRITICAL"
)

// EventCategory classifies operational events at a high level.
type EventCategory string

const (
	EventCategoryProduction  EventCategory = "PRODUCTION"
	EventCategoryQuality     EventCategory = "QUALITY"
	EventCategoryLogistics   EventCategory = "LOGISTICS"
	EventCategoryMaintenance EventCategory = "MAINTENANCE"
	EventCategorySafety      EventCategory = "SAFETY"
	EventCategoryCompliance  EventCategory = "COMPLIANCE"
)

// -- Business Entity --

// EntityType classifies business entities.
type EntityType string

const (
	EntityTypeCustomer EntityType = "CUSTOMER"
	EntityTypeVendor   EntityType = "VENDOR"
	EntityTypePartner  EntityType = "PARTNER"
	EntityTypeInternal EntityType = "INTERNAL"
)

// -- Work Order --

// WorkOrderStatus tracks work order lifecycle.
type WorkOrderStatus string

const (
	WorkOrderStatusDraft      WorkOrderStatus = "DRAFT"
	WorkOrderStatusPending    WorkOrderStatus = "PENDING"
	WorkOrderStatusScheduled  WorkOrderStatus = "SCHEDULED"
	WorkOrderStatusInProgress WorkOrderStatus = "IN_PROGRESS"
	WorkOrderStatusPaused     WorkOrderStatus = "PAUSED"
	WorkOrderStatusComplete   WorkOrderStatus = "COMPLETE"
	WorkOrderStatusCancelled  WorkOrderStatus = "CANCELLED"
)

// WorkOrderPriority indicates work order priority level.
type WorkOrderPriority string

const (
	WorkOrderPriorityLow    WorkOrderPriority = "LOW"
	WorkOrderPriorityNormal WorkOrderPriority = "NORMAL"
	WorkOrderPriorityHigh   WorkOrderPriority = "HIGH"
	WorkOrderPriorityUrgent WorkOrderPriority = "URGENT"
)

// -- Production Order --

// OrderStatus tracks production order lifecycle.
type OrderStatus string

const (
	OrderStatusDraft      OrderStatus = "DRAFT"
	OrderStatusPlanned    OrderStatus = "PLANNED"
	OrderStatusReleased   OrderStatus = "RELEASED"
	OrderStatusInProgress OrderStatus = "IN_PROGRESS"
	OrderStatusPaused     OrderStatus = "PAUSED"
	OrderStatusComplete   OrderStatus = "COMPLETE"
	OrderStatusClosed     OrderStatus = "CLOSED"
	OrderStatusCancelled  OrderStatus = "CANCELLED"
)

// -- Quality --

// SampleOutcome is the result of a quality sample evaluation.
type SampleOutcome string

const (
	SampleOutcomePass         SampleOutcome = "PASS"
	SampleOutcomeFail         SampleOutcome = "FAIL"
	SampleOutcomeInconclusive SampleOutcome = "INCONCLUSIVE"
	SampleOutcomePending      SampleOutcome = "PENDING"
)
